### DISCUSSION TITLE
Downgrade from v4 to v3

### DISCUSSION THREAD

**Discussion by elbill** [globaleaks/globaleaks-whistleblowing-software#4543] [Q&A]
Is it possible to downgrade from v4 to v3? 
In order to see if I can restore the score column functionality (#2831) I tried to downgrade my test system from v4.0.18 to 3.11.70 however this did not work. 
I stopped the service and used apt-get install globaleaks=3.11.70 
Installation went smoothly but could not access the system. When I went back to v4 everything was ok.

**Comment by evilaliv3** [comment]
I'm sorry @elbill at the current time downgrade is not supported.

We have plan to support a backup and restore functionality but when implemented it will work just restoring data at the time of the update.

If you want to achieve this you can just:
- install 3.11.70
- restore a backup copy of your data at the time of 3.11.70

I hope this helps. Please let me know if you have any question.

**Comment by elbill** [comment]
Thanks @evilaliv3

**Comment by elbill** [comment]
@evilaliv3 can you please propose a method for removing globaleaks and installing an older version.
I did some DIY by stopping the process, using remove globaleaks and editing install-globaleaks.sh (replacing globaleaks with globaleaks=3.11.70) but didnt seem to work.

**Comment by evilaliv3** [comment]
@elbill: Are you trying on a new server or on the same server?

If you are trying on the same server you can probably achieve your goal issuing just:
```apt-get install globaleaks=3.11.70```

**Comment by larrykind** [comment]
Hi, 
trying to downgrade from 4.0.18 to 3.11.41, that's what I get from the logs:

`2020-05-06 09:36:48+0200 [-] [E] Found an already initialized database version: 52
2020-05-06 09:36:48+0200 [-] [E] Performing schema migration from version 52 to version 51
2020-05-06 09:36:48+0200 [-] [E] Migration failure: can only concatenate str (not "NoneType") to str
2020-05-06 09:36:48+0200 [-] [E] Verbose exception traceback:
2020-05-06 09:36:48+0200 [-] [I] Traceback (most recent call last):\n\n  File "/usr/lib/python3/dist-packages/globaleaks/db/__init__.py", line 105, in update_db\n    migration.perform_migration(db_version)\n\n  File "/usr/lib/python3/dist-packages/globaleaks/db/migration.py", line 277, in perform_migration\n    perform_data_update(new_db_file)\n\n  File "/usr/lib/python3/dist-packages/globaleaks/db/migration.py", line 125, in perform_data_update\n    session = get_session(make_db_uri(db_file), foreign_keys=False)\n\n  File "/usr/lib/python3/dist-packages/globaleaks/orm.py", line 35, in make_db_uri\n    return prefix + db_file\n\nTypeError: can only concatenate str (not "NoneType") to str\n
2020-05-06 09:36:48+0200 [-] Server Shut Down.
2020-05-06 09:36:48+0200 [-] Exiting GlobaLeaks
`

The only way I found to make the older version to work is to make the older version installer installs a brand new db file, by backing-up the newer database and removing it from /var/globaleaks.
I didn't find the way to preserve data in the database after a such downgrade.

If you re-uprgade evrything is still working, but it seems that db schema downgrade was not performed.

**Comment by elbill** [comment]
@evilaliv3  apt-get install globaleaks=3.11.70 over the v4 will not work.
The old version has to be removed and the install script may have to be modified to make an installation of the older version. Since my issue #2831 with the score column has been resolved there is no need to downgrade any more. @larrykind you may have to see this in more detail if this is important. I confirm your findings.

**Comment by evilaliv3** [comment]
You are correct @elbill , thank you for the clarification.

As prerequisite for downgrading to version X you need a backup copy of your data at the time of version X.
Then you should proceed as following:
- stop globaleaks
- uninstall globaelaks
- replace the existing /var/globaleaks with the backup copy of your data at the time of version X
- apt-get install globaleaks==versionX

This is actually similar to the procedure you need to apply if you want to migrate globaleaks from one server to an other: https://docs.globaleaks.org/en/latest/user/admin/UpgradeGuide.html#upgrade-of-the-distribution-version
